homeassistant/components/freedompro/light.py

```diff
@@ -0,0 +1,114 @@
+"""Support for Freedompro light."""
+import json
+import math
+
+from pyfreedompro import put_state
+
+from homeassistant.components.light import (
+    ATTR_BRIGHTNESS,
+    ATTR_HS_COLOR,
+    COLOR_MODE_BRIGHTNESS,
+    COLOR_MODE_HS,
+    COLOR_MODE_ONOFF,
+    LightEntity,
+)
+from homeassistant.const import CONF_API_KEY
+from homeassistant.core import callback
+from homeassistant.helpers import aiohttp_client
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from .const import DOMAIN
+
+
+async def async_setup_entry(hass, entry, async_add_entities):
+    """Set up Freedompro light."""
+    api_key = entry.data[CONF_API_KEY]
+    coordinator = hass.data[DOMAIN][entry.entry_id]
+    async_add_entities(
+        Device(hass, api_key, device, coordinator)
+        for device in coordinator.data
+        if device["type"] == "lightbulb"
+    )
+
+
+class Device(CoordinatorEntity, LightEntity):
+    """Representation of an Freedompro light."""
+
+    def __init__(self, hass, api_key, device, coordinator):
+        """Initialize the Freedompro light."""
+        super().__init__(coordinator)
+        self._hass = hass
+        self._session = aiohttp_client.async_get_clientsession(self._hass)
+        self._api_key = api_key
+        self._attr_name = device["name"]
+        self._attr_unique_id = device["uid"]
+        self._type = device["type"]
+        self._characteristics = device["characteristics"]
+        self._attr_is_on = False
+        self._attr_brightness = 0
+        color_mode = COLOR_MODE_ONOFF
+        if "hue" in self._characteristics:
+            color_mode = COLOR_MODE_HS
+        elif "brightness" in self._characteristics:
+            color_mode = COLOR_MODE_BRIGHTNESS
+        self._attr_color_mode = color_mode
+        self._attr_supported_color_modes = {color_mode}
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        device = next(
+            (
+                device
+                for device in self.coordinator.data
+                if device["uid"] == self._attr_unique_id
+            ),
+            None,
+        )
+        if device is not None and "state" in device:
+            state = device["state"]
+            if "on" in state:
+                self._attr_is_on = state["on"]
+            if "brightness" in state:
+                self._attr_brightness = math.floor(state["brightness"] / 100 * 255)
+            if "hue" in state and "saturation" in state:
+                self._attr_hs_color = (state["hue"], state["saturation"])
+        super()._handle_coordinator_update()
+
+    async def async_added_to_hass(self) -> None:
+        """When entity is added to hass."""
+        await super().async_added_to_hass()
+        self._handle_coordinator_update()
+
+    async def async_turn_on(self, **kwargs):
+        """Async function to set on to light."""
+        self._attr_is_on = True
```

Comment:
Why is the state updated optimistically here, yet `await self.coordinator.async_request_refresh()` is called below?
Is the API unreliable and may need a long time to update?

Reply:
Actually, the `await self.coordinator.async_request_refresh()` isn't needed, as the entity is updated after a change.

Reply:
So discussing this a bit with @emontnemery 

The question is, is this an optimistic implementation or not? If there is a need to do this optimistic, the `await self.coordinator.async_request_refresh()` could be dropped. If it isn't optimistic, setting the state can be dropped.